contributors/design-proposals/storage/local-volume-management.md

```diff
@@ -0,0 +1,219 @@
+	deviceMountPath, err := volumeDeviceMounter.MountDevice(volumeToMount.VolumeSpec, volumeToMount.DevicePath, volumeToMount.Pod)
+	if err != nil {
+		// On failure, return error. Caller will log and retry.
+		return volumeToMount.GenerateError("MountVolume.MountDevice failed", err)
+	}
+
+	glog.Infof(volumeToMount.GenerateMsgDetailed("MountVolume.MountDevice succeeded", fmt.Sprintf("device mount path %q", deviceMountPath)))
+
+	// Update actual state of world to reflect volume is globally mounted
+	markDeviceMountedErr := actualStateOfWorld.MarkDeviceAsMounted(
+		volumeToMount.VolumeName)
+	if markDeviceMountedErr != nil {
+		// On failure, return error. Caller will log and retry.
+		return volumeToMount.GenerateError("MountVolume.MarkDeviceAsMounted failed", markDeviceMountedErr)
+	}
+}
+```
+
+##### <a name="k8s_vm_umount"></a>Umount Device Operation
+
+similarly, volume manager invokes `operationGenerator.GenerateUnmountDeviceFunc` to umount the volume device from global path. Instead of `AttachableVolumePlugin`, we should use `DeviceMountableVolumePlugin` to do this umount operation, this is a simple 'Name Replace' refactor.
+
+Moreover, volume manager uses `AttachedVolume.DevicePath` to check if the device is opened, however, this field is retrieved from `Node.Status.VolumesAttached`, which is empty for non attachable volume plugins, so we retrieve device path from the global mount path for volumes only implement `DeviceMountableVolumePlugin` before the umount operation, this can be done with the `mount.GetDeviceNameFromMount` method.
+
+It must be noted that, although we keep this validation for volumes only implement `DeviceMountableVolumePlugin`, this is usually a useless check. Because we do this check after the `UnmountDevice` method succeeded, for `DeviceMountableVolumePlugin` volumes, this means the device is already detached from the node, so this validation can always be passed. Consider this, we require the volume plugins should also execute this validation.
+
+#### <a name="k8s_ad_controller_refactor"></a>A/D Controller refactor
+
+A/D controller needn't to change any implementation details as it only handle `AttachableVolumePlugin` and we actually don't modify this interface and according volume plugins. The only thing we need to do is remove local volume plugins from the returned list of `ProbeAttachableVolumePlugins` method as they aren't `AttachableVolumePlugin` any more.
+
+### <a name="volume_plugin_refactor"></a>Volume Plugin Side Refactor
+
+Currently, some local volumes('rbd', `fc`, `iscsi`) also implement `AttachableVolumePlugin`. As these volumes don't have a remote `Attach` API, the local attach operation is executed inside `WaitForAttach` method, the umount and detach operation both performed inside `UnmountDevice` method, this implement leads to the problems mentioned in the beginning of this article. So we should make these volume plugins implementing `DeviceMountableVolumePlugin` only.
```

Comment:
No, empty `Attach()` and `Detach()` make sure that the volume is not attached to multiple nodes. iSCSI, FC and RBD should still use it. IMO, only LocalStorage can benefit from separate `DeviceMountableVolumePlugin`.